File: internal/tui/views/analytics.go
package views

import (
	"context"
	"fmt"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/mailersend/mailersend-cli/internal/sdkclient"
	"github.com/mailersend/mailersend-cli/internal/tui/components"
	"github.com/mailersend/mailersend-cli/internal/tui/theme"
	"github.com/mailersend/mailersend-cli/internal/tui/types"
	"github.com/mailersend/mailersend-go"
)

var (
	statBoxStyle = lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(theme.Muted).
			Padding(0, 2).
			Margin(0, 1)

	statLabelStyle = lipgloss.NewStyle().
			Foreground(theme.Muted)

	statValueStyle = lipgloss.NewStyle().
			Bold(true).
			Foreground(theme.Primary)

	statGoodStyle = lipgloss.NewStyle().
			Bold(true).
			Foreground(theme.Success)

	statBadStyle = lipgloss.NewStyle().
			Bold(true).
			Foreground(theme.Error)
)

// AnalyticsView displays analytics data.
type AnalyticsView struct {
	client    *mailersend.Mailersend
	transport *sdkclient.CLITransport
	table     components.Table
	data      types.AnalyticsData
	loading   bool
	err       error
	width     int
	height    int
	focused   bool
	dateRange string // "7d", "30d", "90d"
}

// NewAnalyticsView creates a new analytics view.
func NewAnalyticsView(client *mailersend.Mailersend, transport *sdkclient.CLITransport) AnalyticsView {
	columns := []components.Column{
		{Title: "DATE", Width: 12},
		{Title: "SENT", Width: 10},
		{Title: "DELIVERED", Width: 10},
		{Title: "OPENS", Width: 10},
		{Title: "CLICKS", Width: 10},
		{Title: "BOUNCED", Width: 10},
	}
	table := components.NewTable(columns)
	table.SetEmptyMessage("No analytics data found.")

	return AnalyticsView{
		client:    client,
		transport: transport,
		table:     table,
		loading:   true,
		dateRange: "7d",
	}
}

// SetClient updates the SDK client.
func (v *AnalyticsView) SetClient(client *mailersend.Mailersend, transport *sdkclient.CLITransport) {
	v.client = client
	v.transport = transport
}

// SetSize sets the view dimensions.
func (v *AnalyticsView) SetSize(width, height int) {
	v.width = width
	v.height = height
	// Reserve space for summary stats
	v.table.SetSize(width, height-6)
}

// SetFocused sets whether this view is focused.
func (v *AnalyticsView) SetFocused(focused bool) {
	v.focused = focused
	v.table.SetFocused(focused)
}

// Loading returns whether the view is loading.
func (v AnalyticsView) Loading() bool {
	return v.loading
}

// Error returns any error.
func (v AnalyticsView) Error() error {
	return v.err
}

// ItemCount returns the number of items.
func (v AnalyticsView) ItemCount() int {
	return len(v.data.Stats)
}

func (v AnalyticsView) daysFromRange() int {
	switch v.dateRange {
	case "30d":
		return 30
	case "90d":
		return 90
	default:
		return 7
	}
}

// Fetch returns a command to fetch analytics.
func (v AnalyticsView) Fetch() tea.Cmd {
	return func() tea.Msg {
		if v.client == nil {
			return types.AnalyticsLoadedMsg{Err: nil}
		}

		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		now := time.Now()
		days := v.daysFromRange()
		dateFrom := now.AddDate(0, 0, -days).Unix()
		dateTo := now.Unix()

		events := []string{"sent", "delivered", "opened", "clicked", "hard_bounced", "soft_bounced"}

		result, _, err := v.client.Analytics.GetActivityByDate(ctx, &mailersend.AnalyticsOptions{
			DateFrom: dateFrom,
			DateTo:   dateTo,
			GroupBy:  "days",
			Event:    events,
		})

		if err != nil {
			return types.AnalyticsLoadedMsg{Err: sdkclient.WrapError(v.transport, err)}
		}

		// Calculate totals
		data := types.AnalyticsData{
			Stats:    result.Data.Stats,
			DateFrom: now.AddDate(0, 0, -days).Format("2006-01-02"),
			DateTo:   now.Format("2006-01-02"),
			GroupBy:  "days",
		}

		for _, stat := range result.Data.Stats {
			data.Sent += stat.Sent
			data.Delivered += stat.Delivered
			data.Opens += stat.Opened
			data.Clicks += stat.Clicked
			data.Bounced += stat.HardBounced + stat.SoftBounced
		}

		return types.AnalyticsLoadedMsg{
			Data: data,
			Err:  nil,
		}
	}
}

// Update handles messages for this view.
func (v AnalyticsView) Update(msg tea.Msg) (AnalyticsView, tea.Cmd) {
	switch msg := msg.(type) {
	case types.AnalyticsLoadedMsg:
		v.loading = false
		v.err = msg.Err
		if msg.Err == nil {
			v.data = msg.Data
			v.updateTable()
		}
	}
	return v, nil
}

// HandleKey handles key events when this view is active.
func (v *AnalyticsView) HandleKey(msg tea.KeyMsg) tea.Cmd {
	switch msg.String() {
	case "j", "down":
		v.table.MoveDown()
	case "k", "up":
		v.table.MoveUp()
	case "g":
		v.table.GotoTop()
	case "G":
		v.table.GotoBottom()
	case "r":
		v.loading = true
		v.table.SetLoading(true)
		return v.Fetch()
	case "1":
		v.dateRange = "7d"
		v.loading = true
		return v.Fetch()
	case "2":
		v.dateRange = "30d"
		v.loading = true
		return v.Fetch()
	case "3":
		v.dateRange = "90d"
		v.loading = true
		return v.Fetch()
	}
	return nil
}

func (v *AnalyticsView) updateTable() {
	var rows [][]string
	for _, stat := range v.data.Stats {
		rows = append(rows, []string{
			stat.Date,
			fmt.Sprintf("%d", stat.Sent),
			fmt.Sprintf("%d", stat.Delivered),
			fmt.Sprintf("%d", stat.Opened),
			fmt.Sprintf("%d", stat.Clicked),
			fmt.Sprintf("%d", stat.HardBounced+stat.SoftBounced),
		})
	}
	v.table.SetRows(rows)
	v.table.SetLoading(false)
}

// View renders the analytics view.
func (v AnalyticsView) View() string {
	var b strings.Builder

	// Summary stats row
	stats := []struct {
		label string
		value int
		style lipgloss.Style
	}{
		{"Sent", v.data.Sent, statValueStyle},
		{"Delivered", v.data.Delivered, statGoodStyle},
		{"Opens", v.data.Opens, statValueStyle},
		{"Clicks", v.data.Clicks, statValueStyle},
		{"Bounced", v.data.Bounced, statBadStyle},
	}

	var boxes []string
	for _, s := range stats {
		box := statBoxStyle.Render(
			statLabelStyle.Render(s.label) + "\n" +
				s.style.Render(fmt.Sprintf("%d", s.value)),
		)
		boxes = append(boxes, box)
	}

	b.WriteString(lipgloss.JoinHorizontal(lipgloss.Top, boxes...))
	b.WriteString("\n\n")

	// Date range indicator
	rangeText := fmt.Sprintf("Date range: %s to %s (%s)", v.data.DateFrom, v.data.DateTo, v.dateRange)
	b.WriteString(lipgloss.NewStyle().Foreground(theme.Muted).Render(rangeText))
	b.WriteString("  ")
	b.WriteString(lipgloss.NewStyle().Foreground(theme.Key).Render("[1]7d [2]30d [3]90d"))
	b.WriteString("\n\n")

	// Table
	b.WriteString(v.table.View())

	return b.String()
}

File: src/tui/views/analytics.rs
use anyhow::Result;
use chrono::{Duration, Utc};
use crossterm::event::KeyCode;
use ratatui::layout::{Constraint, Layout, Rect};
use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::Paragraph;
use ratatui::Frame;
use serde_json::Value;

use super::KeyResult;
use crate::api::Client;
use crate::tui::components::pad_right;
use crate::tui::components::table::{Cell, Column, Table};
use crate::tui::theme;
use crate::util::jint;
use crate::util::jstr;

pub const EMPTY: &str = "No analytics data found.";

const EVENTS: [&str; 6] = [
    "sent",
    "delivered",
    "opened",
    "clicked",
    "hard_bounced",
    "soft_bounced",
];

#[derive(Default)]
pub struct AnalyticsData {
    pub stats: Vec<Value>,
    pub date_from: String,
    pub date_to: String,
    pub sent: i64,
    pub delivered: i64,
    pub opens: i64,
    pub clicks: i64,
    pub bounced: i64,
}

pub fn columns() -> Vec<Column> {
    vec![
        Column {
            title: "DATE",
            width: 12,
        },
        Column {
            title: "SENT",
            width: 10,
        },
        Column {
            title: "DELIVERED",
            width: 10,
        },
        Column {
            title: "OPENS",
            width: 10,
        },
        Column {
            title: "CLICKS",
            width: 10,
        },
        Column {
            title: "BOUNCED",
            width: 10,
        },
    ]
}

pub fn fetch(client: &Client, days: i64) -> Result<AnalyticsData> {
    let now = Utc::now();
    let from = now - Duration::days(days);
    let mut query: Vec<(&str, String)> = vec![
        ("date_from", from.timestamp().to_string()),
        ("date_to", now.timestamp().to_string()),
        ("group_by", "days".to_string()),
    ];
    for e in EVENTS {
        query.push(("event[]", e.to_string()));
    }

    let body = client.get_with("/analytics/date", &query)?;
    let stats: Vec<Value> = body["data"]
        .get("stats")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();

    let mut data = AnalyticsData {
        stats: stats.clone(),
        date_from: from.format("%Y-%m-%d").to_string(),
        date_to: now.format("%Y-%m-%d").to_string(),
        ..Default::default()
    };
    for stat in &stats {
        data.sent += jint(stat, "sent");
        data.delivered += jint(stat, "delivered");
        data.opens += jint(stat, "opened");
        data.clicks += jint(stat, "clicked");
        data.bounced += jint(stat, "hard_bounced") + jint(stat, "soft_bounced");
    }
    Ok(data)
}

pub struct View {
    pub table: Table,
    pub data: AnalyticsData,
    pub loading: bool,
    date_range: &'static str,
}

impl View {
    pub fn new() -> Self {
        let mut table = Table::new(columns(), EMPTY);
        table.set_loading(true);
        Self {
            table,
            data: AnalyticsData::default(),
            loading: true,
            date_range: "7d",
        }
    }

    pub fn days(&self) -> i64 {
        match self.date_range {
            "30d" => 30,
            "90d" => 90,
            _ => 7,
        }
    }

    pub fn set_range(&mut self, range: &'static str) {
        self.date_range = range;
        self.loading = true;
    }

    pub fn set_loaded(&mut self, data: AnalyticsData) {
        let rows = data
            .stats
            .iter()
            .map(|stat| {
                vec![
                    Cell::plain(jstr(stat, "date")),
                    Cell::plain(jint(stat, "sent").to_string()),
                    Cell::plain(jint(stat, "delivered").to_string()),
                    Cell::plain(jint(stat, "opened").to_string()),
                    Cell::plain(jint(stat, "clicked").to_string()),
                    Cell::plain(
                        (jint(stat, "hard_bounced") + jint(stat, "soft_bounced")).to_string(),
                    ),
                ]
            })
            .collect();
        self.data = data;
        self.table.set_rows(rows);
        self.table.set_loading(false);
        self.loading = false;
    }

    pub fn handle_key(&mut self, code: KeyCode) -> KeyResult {
        match code {
            KeyCode::Char('j') | KeyCode::Down => self.table.move_down(),
            KeyCode::Char('k') | KeyCode::Up => self.table.move_up(),
            KeyCode::Char('g') => self.table.goto_top(),
            KeyCode::Char('G') => self.table.goto_bottom(),
            KeyCode::Char('r') => {
                self.loading = true;
                self.table.set_loading(true);
                return KeyResult::Fetch;
            }
            KeyCode::Char('1') => {
                self.set_range("7d");
                return KeyResult::Fetch;
            }
            KeyCode::Char('2') => {
                self.set_range("30d");
                return KeyResult::Fetch;
            }
            KeyCode::Char('3') => {
                self.set_range("90d");
                return KeyResult::Fetch;
            }
            _ => {}
        }
        KeyResult::None
    }

    pub fn render(&mut self, frame: &mut Frame, area: Rect, focused: bool) {
        let [chrome, table_area] =
            Layout::vertical([Constraint::Length(7), Constraint::Min(0)]).areas(area);
        frame.render_widget(Paragraph::new(self.summary_chrome()), chrome);

        self.table.set_focused(focused);
        self.table
            .set_size(table_area.width as usize, table_area.height as usize);
        frame.render_widget(Paragraph::new(self.table.lines()), table_area);
    }

    fn summary_chrome(&self) -> Vec<Line<'static>> {
        let stats: [(&str, String, Color); 5] = [
            ("Sent", self.data.sent.to_string(), theme::PRIMARY),
            ("Delivered", self.data.delivered.to_string(), theme::SUCCESS),
            ("Opens", self.data.opens.to_string(), theme::PRIMARY),
            ("Clicks", self.data.clicks.to_string(), theme::PRIMARY),
            ("Bounced", self.data.bounced.to_string(), theme::ERROR),
        ];

        let mut top: Vec<Span<'static>> = Vec::new();
        let mut label_line: Vec<Span<'static>> = Vec::new();
        let mut value_line: Vec<Span<'static>> = Vec::new();
        let mut bottom: Vec<Span<'static>> = Vec::new();

        for (i, (label, value, color)) in stats.iter().enumerate() {
            if i > 0 {
                top.push(Span::raw(" "));
                label_line.push(Span::raw(" "));
                value_line.push(Span::raw(" "));
                bottom.push(Span::raw(" "));
            }
            let inner = label.chars().count().max(value.chars().count());
            let width = inner + 4;
            let border = Style::new().fg(theme::MUTED);

            top.push(Span::styled(format!("╭{}╮", "─".repeat(width - 2)), border));
            label_line.push(Span::styled(
                format!("│  {}  │", pad_right(label, inner)),
                Style::new().fg(theme::MUTED),
            ));
            value_line.push(Span::styled(
                format!("│  {}  │", pad_right(value, inner)),
                Style::new().fg(*color).add_modifier(Modifier::BOLD),
            ));
            bottom.push(Span::styled(format!("╰{}╯", "─".repeat(width - 2)), border));
        }

        let range = format!(
            "Date range: {} to {} ({})",
            self.data.date_from, self.data.date_to, self.date_range
        );

        vec![
            Line::from(top),
            Line::from(label_line),
            Line::from(value_line),
            Line::from(bottom),
            Line::raw(""),
            Line::from(vec![
                Span::styled(range, Style::new().fg(theme::MUTED)),
                Span::raw("  "),
                Span::styled("[1]7d [2]30d [3]90d", Style::new().fg(theme::KEY)),
            ]),
            Line::raw(""),
        ]
    }
}
